### PR TITLE
feat: add `last-value` decorator

### LINKED ISSUE #12: Decorator for aliasing the last value
I find myself using this pattern a lot in my code:

```javascript
class Foo
  @task
  someTask = function*() { /* whatever */ };

  @alias('someTask.lastSuccessful.value')
  someTaskValue;
}
```

I came up with a little helper that just removes the need to add the `lastSuccessful.value` bit to the end, and allows an option for providing a default value. Would you be interested in adding the helper to this project? Or do you want to keep the decorators to just task definition?

### PR DESCRIPTION
**Pull request by alexlafroscia**
Closes #12

### CODE REVIEW

**Comment by alexlafroscia** [comment]
Not sure if you'll be able to access this page, but the tests are passing now on Travis with the removal of those `ember-try` cases that didn't actually have a configuration:

https://travis-ci.com/alexlafroscia/ember-concurrency-decorators/builds/86374121

**Comment by buschtoens** [comment]
Sorry for letting this sit around so long. The notification must have slipped past me. 

Looking great already! What do you think about using an initializer instead of a `default` / `defaultFactory` option?

```js
@lastValue('someTask')
value = [];
```

The cool thing is, that `desc.initializer` already is a (factory) function.

**Comment by alexlafroscia** [comment]
Totally, that makes way more sense!

Should the semantics around the "default" changing slightly then? Right now it's "use the default (or factory) if the task has not been run _or_ if the task returned `undefined` or `null`".  If we're using the initializer instead, it might make more sense for it _only_ to use that value if the task hasn't finished yet.

**Comment by buschtoens** [comment]
Maybe it's personal preference, but I always write tasks in a way that they either return a value that can be handled by the receiver or else throw an error. From that POV I would expect the decorator to only return the default value, if there is no `lastSuccessful` yet.

_However_ this might just be me / us. We could add a `defaultIfNone` / `defaultIfNullish` (naming things is hard, let's bikeshed :D ) option:

```js
@lastValue('someTask', { defaultIfNone: true })
value = [];
```

**Comment by buschtoens** [comment]
A second thing: I think it makes total sense that `@lastValue` reads `lastSucessful.value`, as in most cases you only care about successful tasks. But we could (potentially in a later PR, so we can :shipit: now) add another option like:

```js
@lastValue('someTask', { onlySuccessful: false })
value = [];
```

**Comment by alexlafroscia** [comment]
> I always write tasks in a way that they either return a value that can be handled by the receiver or else throw an error

Same here; I think that's reasonable advice to give to people using `ember-concurrency` as well.

I think we should start with the most basic API, like you suggested initially, where the initializer provides the default value. We can expand the API with more options and stuff later on, but it may not be necessary. I'm definitely of the mind to keep it simple early on and let people request features 😛

File: addon/last-value.js [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+ * @function
+ * @param {string} taskName the name of the task to read a value from
+ */
+export default computedDecoratorWithParams(function(target, key, desc, [task]) {
+  const taskValueKey = `${task}.lastSuccessful.value`;
+
+  return computed(taskValueKey, function() {
+    if (get(this, `${task}.lastSuccessful`)) {
+      return get(this, taskValueKey);
+    }
```
Comment (buschtoens) [inline addon/last-value.js]:
```js
    const lastSuccessful = get(this, `${task}.lastSuccessful`);
    if (lastSuccesful) {
      return get(lastSuccesful, 'value');
    }
```

is probably a μs faster.

File: addon-test-support/in-run-loop.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
+import { begin, end } from "@ember/runloop";
+
+/**
+ * Ensures that each test is executed within a RunLoop
+ *
+ * @param {object} hooks QUnit Hooks
+ */
+export default function inRunloop(hooks) {
```
Comment (buschtoens) [inline addon-test-support/in-run-loop.js]: Nice!

**Comment by buschtoens** [comment]
Out of curiosity: what auto-formatter are you using? Prettier?

**Comment by alexlafroscia** [comment]
Yeah, prettier. Sorry about that, I can remove the bits it changed in the README. 

Or, if you’d like, I’m happy to PR a Prettier config that ties into ESLint if you want to go that direction.

**Comment by buschtoens** [comment]
No need to be sorry. 😊 

> I’m happy to PR a Prettier config that ties into ESLint if you want to go that direction.

That would be extra sweet! Didn't get around to it yet unfortunately. Can you please set `singleQuotes` then?

**Comment by alexlafroscia** [comment]
Just re-based the PR to fix your minor performance note and fix the `yarn.lock` merge conflict

File: addon/last-value.js (hunk 1 above)
Comment (alexlafroscia) [reply addon/last-value.js]: Fixed!

File: addon-test-support/in-run-loop.js (hunk 2 above)
Comment (alexlafroscia) [reply addon-test-support/in-run-loop.js]: Yeah, I found this was necessary to make the tests behave more like a user's actual application would, where the intermediary run-loop needing to start will just happen automatically.

A nice bonus of putting this in `addon-test-support` is that people can import it and use it themselves if they want!

File: package.json [hunk 3]
```diff
@@ -35,7 +35,6 @@
     "broccoli-asset-rev": "^3.0.0",
     "ember-cli": "~3.3.0",
     "ember-cli-dependency-checker": "^3.0.0",
-    "ember-cli-eslint": "^4.2.1",
```
Comment (alexlafroscia) [inline package.json]: Note: `eslint-plugin-prettier` doesn't play nicely with `ember-cli-eslint`. I think this is OK though -- running linting in your unit tests isn't really necessary so I think it's a fair trade-off for an easier `prettier` experience.

**Comment by alexlafroscia** [comment]
Passing build can be seen on my fork here:

https://travis-ci.com/alexlafroscia/ember-concurrency-decorators/builds/87582000?utm_medium=notification&utm_source=email

If you're interested in adding a maintain, I'd be happy to get Travis configured for you on this repo!

**Comment by alexlafroscia** [comment]
I also have a PR ready for when this one is merged that runs `ember-cli-update` and updates a bunch of dependencies!

I want to make that as a separate PR after this lands because there would have been conflicts between this and that branch and it was easier to just update from this branch as the starting point.

File: package.json (hunk 3 above)
Comment (buschtoens) [reply package.json]: I share this opinion 😄 

https://github.com/ember-cli/rfcs/pull/121

**Comment by buschtoens** [comment]
> If you're interested in adding a maintain, I'd be happy to get Travis configured for you on this repo! It can integration with the new "Checks" tab on pull requests, it's pretty sweet.

Yes, I'd love to have you onboard! Thank you for offering it. I've pinged @machty, so that he can add you. 🙂

**Pull request merged**